cosmic-core/server/src/main/java/com/cloud/network/NetworkServiceImpl.java

```diff
@@ -996,6 +1001,45 @@ public Network createGuestNetwork(final CreateNetworkCmd cmd) throws Insufficien
         return network;
     }
 
+    public void checkIpExclusionList(final String ipExclusionList, final String cidr, List<NicVO> nicsPresent) {
+        if (!org.apache.commons.lang.StringUtils.isEmpty(ipExclusionList)) {
+            // validate ipExclusionList
+            // Perform a "syntax" check on the list
+            if (!NetUtils.validIpRangeList(ipExclusionList)) {
+                throw new InvalidParameterValueException("Syntax error in ipExclusionList");
+            }
+
+            final List<String> excludedIps = NetUtils.getAllIpsFromRangeList(ipExclusionList);
+
+            if (cidr != null) {
+                //Check that ipExclusionList (delimiters) is within the CIDR
+                if(!NetUtils.isIpRangeListInCidr(ipExclusionList,cidr)){
+                    throw new InvalidParameterValueException("An IP in the ipExclusionList is not part of the CIDR of the network " + cidr);
+                }
+
+                //Check that at least one IP is available after exclusion for the router interface
+                final SortedSet<Long> allPossibleIps = NetUtils.getAllIpsFromCidr(cidr, NetUtils.listIp2LongList(excludedIps));
+                if( allPossibleIps.size() == 0 ){
+                    throw new InvalidParameterValueException("The ipExclusionList excludes all IPs in the CIDR; at least one needs to be available");
+                }
+            }
+
+            if (nicsPresent != null) {
+                // Check that no existing nics/ips are part of the exclusion list
+                for (final NicVO nic : nicsPresent) {
+                    final String nicIp = nic.getIPv4Address();
+                    //check if nic IP is exclusionList
+                    if (excludedIps.contains(nicIp)) {
```

Comment:
These 2 `if`s can be combined: `if (excludedIps.contains(nicIp) && !Nic.State.Deallocating.equals(nic.getState())) { throw ...`